cumulusci/tasks/bulkdata/step.py

```diff
@@ -0,0 +1,246 @@
+
+
+class BulkApiQueryStep(QueryStep, BulkJobTaskMixin):
+    def query(self):
+        self.job_id = self.bulk.create_query_job(self.sobject, contentType="CSV")
+        self.batch_id = self.bulk.query(self.job_id, self.soql)
+
+        result = self._wait_for_job(self.job_id)
+        if result == "Completed":
+            self.status = Status.SUCCESS
+        else:
+            self.status = Status.FAILURE
+
+        self.bulk.close_job(self.job_id)
+
+    def get_results(self):
+        # FIXME: For PK Chunking, need to get new batch Ids
+        # and retrieve their results. Original batch will not be processed.
+
+        result_ids = self.bulk.get_query_batch_result_ids(
+            self.batch_id, job_id=self.job_id
+        )
+        for result_id in result_ids:
+            uri = f"{self.bulk.endpoint}/job/{self.job_id}/batch/{self.batch_id}/result/{result_id}"
+
+            with download_file(uri, self.bulk) as f:
+                reader = csv.reader(f)
+                self.headers = next(reader)
+                if "Records not found for this query" in self.headers:
+                    return
+
+                yield from reader
+
+
+class DmlStep(Step):
+    def __init__(self, sobject, operation, api_options, context, fields):
+        super().__init__(sobject, operation, api_options, context)
+        self.fields = fields
+
+    def start(self):
+        pass
+
+    def load_records(self, records):
+        pass
+
+    def end(self):
+        pass
+
+    def get_results(self):
+        return []
+
+
+class BulkApiDmlStep(DmlStep, BulkJobTaskMixin):
+    def start(self):
+        self.job_id = self.bulk.create_job(
+            self.sobject,
+            self.operation.value,
+            contentType="CSV",
+            concurrency=self.api_options.get("bulk_mode", "Parallel"),
+        )
+
+    def end(self):
+        self.bulk.close_job(self.job_id)
+        result = self._wait_for_job(self.job_id)
+        if result == "Completed":
+            self.status = Status.SUCCESS
+        else:
+            self.status = Status.FAILURE
+
+    def load_records(self, records):
+        self.batch_ids = []
+
+        for count, batch_file in enumerate(self._batch(records)):
+            self.context.logger.info(f"Uploading batch {count + 1}")
+            self.batch_ids.append(self.bulk.post_batch(self.job_id, batch_file))
+
+    def _batch(self, records):
+        for batch in BatchIterator(records, self.api_options.get("batch_size", 10000)):
+            yield self._csv_generator(batch)
+
+    def _csv_generator(self, records):
+        content = io.StringIO()
+        writer = csv.writer(content)
+        writer.writerow(self.fields)
+
+        content.seek(0)
+        yield content.read().encode("utf-8")
+        for rec in records:
+            content = io.StringIO()
+            writer = csv.writer(content)
+            writer.writerow(rec)
+            content.seek(0)
+
+            yield content.read().encode("utf-8")
```

Comment:
It seems like there should be a way to do this without constructing a new csv writer for each row.

Reply:
I did some experiments and sure enough, it's considerably faster to reuse the `csv.writer` and `StringIO` instances (truncating the `StringIO` each row) than it is to recreate them. Fixing!